test/acceptance/tests/hcp/hcp_test.go

```diff
@@ -155,6 +150,402 @@ func TestHCP(t *testing.T) {
+			}},
+		}, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func deleteExportedServices(t *testing.T, consulClient *api.Client, dst *helpers.MeshTask) {
+	retry.RunWith(timeouts["consul"], t, func(r *retry.R) {
+		_, err := consulClient.ConfigEntries().Delete(api.ExportedServices, dst.Partition, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func consulClient(t *testing.T, addr, token string) (*api.Client, ConsulState, error) {
+	cfg := api.DefaultConfig()
+	cfg.Address = addr
+	cfg.Token = token
+	client, err := api.NewClient(cfg)
+	if err != nil {
+		return nil, ConsulState{}, fmt.Errorf("failed to create Consul client: %w", err)
+	}
+	state, err := recordConsulState(t, client)
+	if err != nil {
+		return nil, ConsulState{}, fmt.Errorf("failed to establish Consul state: %w", err)
+	}
+	return client, state, err
+}
+
+type ConsulState struct {
+	Partitions map[string]PartitionState
+}
+
+type PartitionState struct {
+	Name       string
+	Namespaces map[string]NamespaceState
+}
+
+type NamespaceState struct {
+	Name        string
+	Tokens      map[string]struct{}
+	Policies    map[string]struct{}
+	Roles       map[string]struct{}
+	AuthMethods map[string]struct{}
+}
+
+func recordConsulState(t *testing.T, consul *api.Client) (ConsulState, error) {
+	// TODO: not sure how to handle OSS vs Enterprise cases.. these tests currently use Enterprise.
+	parts, _, err := consul.Partitions().List(context.Background(), nil)
+	if err != nil {
+		return ConsulState{}, err
+	}
+
+	state := ConsulState{Partitions: make(map[string]PartitionState)}
+	for _, part := range parts {
+		t.Logf("recording partition state for %s", part.Name)
+		nss, _, err := consul.Namespaces().List(&api.QueryOptions{Partition: part.Name})
+		if err != nil {
+			return ConsulState{}, err
+		}
+		partState := PartitionState{Name: part.Name, Namespaces: make(map[string]NamespaceState)}
+		for _, ns := range nss {
+
+			t.Logf("  recording namespace state for %s/%s", part.Name, ns.Name)
+			opts := &api.QueryOptions{Partition: part.Name, Namespace: ns.Name}
+			nsState := NamespaceState{Name: ns.Name}
+
+			// Tokens
+			nsState.Tokens = map[string]struct{}{}
+			tokens, _, err := consul.ACL().TokenList(opts)
+			if err != nil {
+				return ConsulState{}, err
+			}
+			for _, tok := range tokens {
+				t.Logf("    recording token %s in %s/%s", tok.AccessorID, part.Name, ns.Name)
+				nsState.Tokens[tok.AccessorID] = struct{}{}
+			}
+
+			// Policies
+			nsState.Policies = map[string]struct{}{}
+			policies, _, err := consul.ACL().PolicyList(opts)
+			if err != nil {
+				return ConsulState{}, err
+			}
+			for _, p := range policies {
+				t.Logf("    recording policy %s (%s) in %s/%s", p.Name, p.ID, part.Name, ns.Name)
+				nsState.Policies[p.ID] = struct{}{}
+			}
+
+			partState.Namespaces[ns.Name] = nsState
+		}
+		state.Partitions[part.Name] = partState
+	}
+	return state, nil
+}
+
+func restoreConsulState(t *testing.T, consul *api.Client, state ConsulState) error {
```

Comment:
I think this is fine, but to simplify, is there a reason we shouldn't delete everything in the HCP cluster? We could clean out the HCP cluster to ensure a clean slate at the beginning of each test case.

I guess we couldn't parallelize easily if we do that (although, we maybe can't parallelize these easily either when the test cases share a server cluster?).

Reply:
Parallel == Hard :laughing: 

I've left this as is for now. I've create a task in the backlog to revisit the acceptance test performance.